# Level2_design: Bitmanipulation Coprocessor Design Verification

The verification environment is setup using [Vyoma's UpTickPro](https://vyomasystems.com) provided for the hackathon.

![](https://i.imgur.com/miWGA1o.png)

## Verification Environment

The [CoCoTb](https://www.cocotb.org/) based Python test is developed as explained. The test drives inputs to the Design Under Test (adder module here) which takes in 32-bit inputs *mav_putvalue_src1* , *mav_putvalue_src2* and *mav_putvalue_src3*. A random number is generated and stored to the inputs. The 32-bit output *mav_putvalue* is generated by passing the inputs to bitmanip() function.

The errors list is created to store in case any bugs are found in design. The instr list contains instruction values to be given to mav_putvalue_instr.
The values are assigned to the input port using 
```
dut.mav_putvalue_src1.value = mav_putvalue_src1
dut.mav_putvalue_src2.value = mav_putvalue_src2
dut.mav_putvalue_src3.value = mav_putvalue_src3
dut.EN_mav_putvalue.value = 1
dut.mav_putvalue_instr.value = mav_putvalue_instr
```

The assert statement is used for comparing the dut_output to the expected output and it is given inside try block. If there is an assertion error,error statement inside except block is printed and value is appended to 'errors' list.

The assert statement outside the for loop checks if there's any bugs appended to 'errors'. If len(errors) is not equal to zero, the whole test fails. The following error is seen:

The following error is seen:
```
assert dut.sum.value == A+B, "Adder result is incorrect: {A} + {B} != {SUM}, expected value={EXP}".format(
                     AssertionError: Adder result is incorrect: 7 + 5 != 2, expected value=12
```
## Test Scenario 
- Test Inputs: mav_putvalue_instr=40007033
- Expected Output: sum=12
- Observed Output in the DUT dut.sum=2

Output mismatches for the above inputs proving that there is a design bug

## Verification Strategy

## Is the verification complete ?
